### DISCUSSION TITLE
Load URL In app

### DISCUSSION THREAD

**Discussion by AnonymousXC** [wailsapp/wails#1138] [Q&A] [answered]
# Load Url in a Wails app
Hi,
Is there a way to load an Url in a wails app.
As using iframe or embed compromises security so, the X-HEADER is enabled.
Thanks!👍⚡⚡

**Comment by redraskal** [comment, marked as answer]
> # Load Url in a Wails app
> Hi, Is there a way to load an Url in a wails app. As using iframe or embed compromises security so, the X-HEADER is enabled. Thanks!👍⚡⚡

## Possible Solution
Modify "devserverurl" in wails.json to the URL you want

**Issue:** Wailsjs is bypassed
##
If you aren't looking to navigate to an external page, maybe use a CORS proxy to bypass the issue? You could use a public one online or launch a local one with the app.
<s>I think I have a similar issue. I'm looking to load an external app via HTTPS and inject JS into the page to modify it (this is technically possible with webview itself but not with wails v2 afaik). My goal is to render some storage website such as Google Drive for instance but include a better download utility than the awful chrome downloading utility that can't handle resuming large file downloads.

Seems like v2 uses JS injection for the wails runtime on build while v1 used JS execution with webview.</s>
Found a solution to my problem.

**Reply by leaanthony** [reply]
Would you mind sharing so others could benefit? 👍

**Reply by redraskal** [reply]
> Would you mind sharing so others could benefit? 👍

So I saw that [ExecJS](https://github.com/wailsapp/wails/blob/c2fa4b6103de28d4ae5988ddae111894e12fd620/v2/internal/frontend/desktop/windows/frontend.go#L477) is a public function on Frontend but its in the internal package. This is the function I was looking for.

I dug around most of v2 code and found a [method](https://github.com/wailsapp/wails/blob/7bc51725c05bd712014aa3fce26807ebaa69e52c/v2/pkg/runtime/runtime.go#L14) to return Frontend from a context but its also private
**(can we make this function public? its not in the internal package)**

The context that includes "frontend" is passed to the app event OnStartup so I created an interface that mirrored the function I wanted and used a type assertion on the context value.

Today I'm going to look at injecting wailsjs manually because using your own URL for devserverurl does not inject wailsjs (since it seems to be injected into the files instead on builds now)

I think it might make sense to rename devserverurl to something more specific though? I wasn't fully sure what it meant until I saw that this is the URL used instead of a local index.html both in development & production.

```go
type App struct {
	ctx      context.Context
	frontend frontend
}

type frontend interface {
	ExecJS(js string)
}

func NewApp() *App {
	return &App{}
}

func (b *App) startup(ctx context.Context) {
	b.ctx = ctx
	if f := ctx.Value("frontend"); f != nil {
		b.frontend = f.(frontend)
	}
}

func (b *App) domReady(ctx context.Context) {
	b.frontend.ExecJS("alert('hi!')")
}
```

**Reply by AnonymousXC** [reply]
> > # Load Url in a Wails app
> > Hi, Is there a way to load an Url in a wails app. As using iframe or embed compromises security so, the X-HEADER is enabled. Thanks!👍⚡⚡
> 
> ## Possible Solution
> 
> Modify "devserverurl" in wails.json to the URL you want
> 
> **Issue:** Wailsjs is bypassed
> 
> If you aren't looking to navigate to an external page, maybe use a CORS proxy to bypass the issue? You could use a public one online or launch a local one with the app. I think I have a similar issue. I'm looking to load an external app via HTTPS and inject JS into the page to modify it (this is technically possible with webview itself but not with wails v2 afaik). My goal is to render some storage website such as Google Drive for instance but include a better download utility than the awful chrome downloading utility that can't handle resuming large file downloads.
> 
> Seems like v2 uses JS injection for the wails runtime on build while v1 used JS execution with webview. Found a solution to my problem.

Thanks for Suggestion.

**Reply by Tazeg** [reply]
In the JS frontend, you can use : `window.location.href='https://example.com'`

**Reply by AnonymousXC** [reply]
> In the JS frontend, you can use : `window.location.href='https://example.com'`

Thanks.
Gonna try it.😀

**Reply by xieting1990** [reply]
It doesn't seem to work,
websocket would be disconnected 
`DEB | [DevWebServer] Websocket client 0x14000094510 connected
DEB | [DevWebServer] Websocket client 0x14000094510 disconnected
`
